lom/src/engine/context.go

```diff
@@ -0,0 +1,394 @@
+                delete (cl.Actions, actName)
+            }
+            delete (p.activeActions, actName)
+            GetSeqHandler().DropRequest(actName)
+        }
+    }
+}
+
+func (p *ClientRegistrations_t) NotifyHeartbeats(actName string,
+            ts int64) {
+    if _, ok := p.activeActions[actName]; ok {
+        p.heartbeatCh <- actName
+    }
+}
+
+type HB_t struct {
+    Actions     []string
+    Timestamp   int64
+}
+
+type HBData_t struct {
+    LoM_Heartbeat HB_t
+}
+
+
+func (p *ClientRegistrations_t) PublishHeartbeats() {
+    /* Map will help combine duplicate heartbeats into one */
+    lst := make(map[string]struct{})
+
+    for {
+        /* Read inside the loop to help refresh any change */
+        hb := GetConfigMgr().GetGlobalCfgInt(ENGINE_HB_INTERVAL_SECS)
+        if hb == 0 {
+            hb = MIN_HB_INTERVAL_SECS
+        }
+        select {
+        case a := <- p.heartbeatCh:
+            lst[a] = struct{}{}
+            /* Collect actions */
+
+        case <- time.After(time.Duration(hb) * time.Second):
+            hb := &HBData_t { HB_t { make([]string, len(lst)), time.Now().Unix()}}
+
+            if len(lst) > 0 {
+                /* Publish collected actions, which could be empty */
+                i := 0
+                for k, _ := range lst {
+                    hb.LoM_Heartbeat.Actions[i] = k
+                    i++
+                }
+                /* Reset collected. */
+                lst = make(map[string]struct{})
+            }
+            PublishEvent(hb)
+        }
+    }
+}
+
+
+/* Request to be sent to client as response to client's recvServerRequest */
+func (p *ClientRegistrations_t) AddServerRequest(
+            actionName string, req *ServerRequestData) error {
+    if (len(actionName) == 0) || (req == nil) {
+        return LogError("Internal error: Nil args (%v) (%v)", actionName, req)
+    } else if a, ok := p.activeActions[actionName]; !ok {
+        return LogError("(%s): Action is not registered yet", actionName)
+    } else if cl, ok := p.activeClients[a.Client]; !ok {
+        return LogError("Internal error: client(%s) for action (%s) not found",
+                a.Client, actionName)
+    } else {
+        if len(cl.pendingWriteRequests) >= cap(cl.pendingWriteRequests) {
+            return LogError("Internal error: pendingWriteRequests is full (%d)",
+                    len(cl.pendingWriteRequests))
+        }
+        cl.pendingWriteRequests <- req
+        return nil
+    }
+}
+
+/* Client's request to read server request via recvServerRequest */
+func (p *ClientRegistrations_t) PendServerRequest(req *LoMRequestInt) error {
+    if req == nil {
+        return LogError("Internal error: Nil req")
+    } else if cl, ok := p.activeClients[req.Req.Client]; !ok {
+        return LogError("Internal error: client(%s) not found", req.Req.Client)
+    } else if len(cl.pendingReadRequests) >= cap(cl.pendingReadRequests) {
```

Comment:
should it be only == ? len(cl.pendingReadRequests) == cap(cl.pendingReadRequests)

Reply:
Yes ">" is not possible. 
Just I tend to put it.
Anyhow fixed it.